### PR TITLE
Issue 5261: Support AggregationFunctions with multiple arguments (contd).

### PR DESCRIPTION
**Pull request by mayankshriv**
This PR is a continuation of https://github.com/apache/incubator-pinot/pull/5259
to address the issue https://github.com/apache/incubator-pinot/issues/5261.

1. Added new field in request.thrift `aggregationFunctionArgs` as a list of String
   arguments for the aggregation funciton.
   - Could not use the existing `aggregationParams` as it is a Map, and functions with
     variable arguments may not provide a name for the arg (to be used as key in Map).
   - Maintain backward compatibility by first check for the new field, and fall back to
     the existing one if it does not exist.

2. Ensure that all calls to the old AggregationInfo.getAggregationParams() is replaced
   with backward compatible AgguregationFunctionUtils.getAggregationArgs().

3. Since most aggregation functions today have just one argument, added a separate api
   AggregationFuncitonContext.getFirstArgument() as an optimization.

4. Cleaned up getColumnName() and getResultColumnName() api's in AggregationFunctionContext
   class to not require the column name argument, as this is already stored in the
   AggregationFunction.

5. Modified all tests to use aggregationFunctionArgs instead of aggregationParams.

TODO:
Remove the AggregationFunctionContext class as AggregationFunctions now store their arguments,
and this class no longer provides any additional value.

### CODE REVIEW

File: pinot-common/src/thrift/request.thrift [hunk 1]
```diff
@@ -102,6 +102,11 @@ struct AggregationInfo {
   2: optional map<string,string> aggregationParams;
   3: optional bool isInSelectList;
 
+  // Backward compatible change to allow aggregation functions to take multiple arguments.
+  // We could not reuse aggregationParams, as it requires argument name (as key), which may not be
+  // available for aggregation functions with variable arguments.
+  4: optional list<string> aggregationFunctionArgs;
```
Comment (kishoreg) [inline pinot-common/src/thrift/request.thrift:108]: too long args/argList/params/paramList/operands

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/AggregationFunctionContext.java [hunk 2]
```diff
@@ -18,19 +18,28 @@
...
+import com.google.common.base.Preconditions;
+import java.util.List;
 import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
 
 
 /**
  * This class caches miscellaneous data to perform efficient aggregation.
+ *
+ * TODO: Remove this class, as it no longer provides any value after aggregation functions now store
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/AggregationFunctionContext.java:30]: +1

Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/AggregationFunctionContext.java:30]: why not remove it in this PR itself?

File: pinot-common/src/main/java/org/apache/pinot/parsers/utils/BrokerRequestComparisonUtils.java [hunk 3]
```diff
@@ -89,8 +89,7 @@ public static boolean validate(BrokerRequest br1, BrokerRequest br2, boolean ign
           return false;
         }
       } else if (br2.getGroupBy() != null) {
-        LOGGER.error("tGroupBy did not match, br1.getGroupBy() = null, br2.getGroupBy() = {}",
-            br2.getGroupBy());
+        LOGGER.error("tGroupBy did not match, br1.getGroupBy() = null, br2.getGroupBy() = {}", br2.getGroupBy());
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-common/src/main/java/org/apache/pinot/parsers/utils/BrokerRequestComparisonUtils.java:92]: Was in the original code that's why I missed it. Will fix.

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java [hunk 4]
```diff
@@ -233,32 +232,32 @@ private AggregationInfo buildAggregationInfo(Function function) {
...
-      for (int i = 0; i < numOperands; i++) {
-        Expression expression = operands.get(i);
-        String columnExpression = getColumnExpression(expression);
-        if (expressions.add(columnExpression)) {
-          // deduplicate the columns
-          if (i != 0) {
-            sb.append(CompilerConstants.AGGREGATION_FUNCTION_ARG_SEPARATOR);
+      // Need to de-dup columns for distinct.
+      if (functionName.equalsIgnoreCase(AggregationFunctionType.DISTINCT.getName())) {
+        Set<String> expressionSet = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (mayankshriv) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java:243]: 👍

File: pinot-common/src/thrift/request.thrift (hunk 1 above)
Comment (mayankshriv) [reply pinot-common/src/thrift/request.thrift:108]: Replaced with `expressions` based on offline discussion.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/AggregationFunctionContext.java (hunk 2 above)
Comment (mayankshriv) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/AggregationFunctionContext.java:30]: Seemed like a bit non-trivial, will address in the next PR.

**Pull request merged**